kernel-headers/CMakeLists.txt

```diff
@@ -0,0 +1,27 @@
+publish_internal_headers(rdma
+  rdma/bnxt_re-abi.h
+  rdma/cxgb3-abi.h
+  rdma/cxgb4-abi.h
+  rdma/hns-abi.h
+  rdma/ib_user_cm.h
+  rdma/ib_user_ioctl_verbs.h
+  rdma/ib_user_mad.h
+  rdma/ib_user_sa.h
+  rdma/ib_user_verbs.h
+  rdma/mlx4-abi.h
+  rdma/mlx5-abi.h
```

Comment:
Yes, I am aware of the extent of the mess.

I intend to slowly over time fix the drivers too, as I already fixed the core code. As that happens the drivers will learn to include the uapi header via something like

```c
#include <rdma/mlxX-abih.>
```

in their abi.h header, just as I did in verbs. The mess is extensive, this will take time.

I prefer to have all the headers files as it makes it less work to keep them in sync, less work to do the above rework, and it runs the headers through our CI checker which is finding bugs in the headers..

Reply:
I'm not sure that people would like to see H files that are not in use in rdma-core, just for future purposes, it's very confusing. The alternative is to take for now only H files that are really in use.  In addition, what will be the methodology to update them ? once per rdma-core release ? when /who will handle that ?  

The question is also open for files that are in use (e.g. the verbs uAPI) we should define a clear definition when those should be updated. This should be done manually I believe during submission of new features to let them compile before the kernel part was accepted and maybe at the end of a release we need some mechanism/script to overwrite  the full file by the latest kernel code assuming that there is no conflict ...

Reply:
It is too complicated to try and figure out what headers are needed or not. This is simple, take the entire directory from the kernel. I don't see why anyone would be confused by this.

Regarding merging patches, what I want to see now is any change to kernel_headers/ be in a single commit that touches no other files. The commit must refer to an upstream commit kernel ID that the headers came from. The commit should sync the entire kernel_headers/ dir with the referenced kernel commit. 

This is similar to what iproute does.

I *expressly* do not want to see any manual updates or partial patches, or *anything* that could result in something other than a byte-for-byte copy of the entire tree from an upstream kernel.

People sending PRs that depend on kernel changes will have to  manage this and update their PRs after the kernel patches are merged.

In addition to that, we may wish to sync kernel headers once per major release or something.